### PR TITLE
Add password expiration strategy

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Add a password expiration mechanism. It can be configured on a new `Security` tab of global configuration. Existing password security policy configuration has been moved in this tab.

image

### CODE REVIEW

File: inc/config.class.php [hunk 1]
```diff
@@ -3475,6 +3480,206 @@ static function showLogsInterval($name, $value, $options = []) {
...
+      echo '<div class="center" id="tabsbody">';
+      echo '<form name="form" action="' . Toolbox::getItemTypeFormURL(__CLASS__) . '" method="post">';
+      echo '<table class="tab_cadre_fixe">';
+      echo '<tr><th colspan="4">' . __('Security setup') . '</th></tr>';
+
+      echo '<tr class="tab_bg_1">';
+      echo '<td colspan="4" class="center b">' . __('Password security policy') . '</td>';
+      echo '</tr>';
+
+      echo '<tr class="tab_bg_2">';
```
Comment (trasher) [inline inc/config.class.php]: This field (among others) is present in two different forms; should not it be removed from its original location?

File: inc/html.class.php [hunk 2]
```diff
@@ -1788,6 +1788,19 @@ static function simpleHeader($title, $links = []) {
 
       // Preferences + logout link
       echo "<div id='c_preference'>";
+      if (Session::getLoginUserID()) {
```
Comment (trasher) [inline inc/html.class.php]: I do not understand why you add a new logout link here

File: inc/user.class.php [hunk 3]
```diff
@@ -4513,6 +4534,92 @@ static function getIdByField($field, $value) {
...
 
 
+   /**
+    * Show password update form for current user.
+    *
+    * @param array $error_messages
+    *
+    * @return void
+    */
+   public function showPasswordUpdateForm(array $error_messages = []) {
```
Comment (trasher) [inline inc/user.class.php]: Original password fields are still present; I guess you should either remove them from original location, or factorize.

File: tests/functionnal/NotificationTargetUser.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+use \DbTestCase;
+
+/* Test for inc/notificationtargetuser.class.php */
+
+class NotificationTargetUser extends DbTestCase {
+
+   public function testAddDataForTemplate() {
+
+      // TODO
```
Comment (trasher) [inline tests/functionnal/NotificationTargetUser.php]: :)

File: inc/html.class.php (hunk 2 above)
Comment (cedric-anne) [reply inc/html.class.php]: This simple header is shown one the password update page which is displayed when a user logs in with an expired password. As this is the only page that user can access, we have to add this logout link to give hi ability to close its session.

Comment (trasher) [reply inc/html.class.php]: Ok :)

File: inc/config.class.php (hunk 1 above)
Comment (cedric-anne) [reply inc/config.class.php]: Yes, it had to be removed. It is done now.

File: inc/user.class.php (hunk 3 above)
Comment (trasher) [reply inc/user.class.php]: As discussed, forms are differents; t is ok to keep as is

File: inc/user.class.php [hunk 5]
```diff
@@ -4513,6 +4534,92 @@ static function getIdByField($field, $value) {
...
+      echo '<input type="password" id="current_password" name="current_password" autocomplete="off" />';
+      echo '</td>';
+      echo '</tr>';
+
+      echo '<tr class="tab_bg_1">';
+      echo '<td>';
+      echo '<label for="password">' . __('New password') . '</label>';
+      echo '</td>';
+      echo '<td>';
+      echo '<input type="password" id="password" name="password" autocomplete="off" onkeyup="return passwordCheck();" />';
```
Comment (orthagh) [inline inc/user.class.php]:
```suggestion
      echo '<input type="password" id="password" name="password" autocomplete="new-password" onkeyup="return passwordCheck();" />';
```

File: inc/user.class.php [hunk 6]
```diff
@@ -4513,6 +4534,92 @@ static function getIdByField($field, $value) {
...
+      echo '</td>';
+      echo '</tr>';
+
+      echo '<tr class="tab_bg_1">';
+      echo '<td>';
+      echo '<label for="password2">' . __('New password confirmation') . '</label>';
+      echo '</td>';
+      echo '<td>';
+      echo '<input type="password" id="password2" name="password2" autocomplete="off" />';
```
Comment (orthagh) [inline inc/user.class.php]:
```suggestion
      echo '<input type="password" id="password2" name="password2" autocomplete="new-password" />';
```

File: tests/functionnal/NotificationTargetUser.php (hunk 4 above)
Comment (cedric-anne) [reply tests/functionnal/NotificationTargetUser.php]: done

**Pull request merged**